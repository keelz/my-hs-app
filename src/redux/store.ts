import { createBrowserHistory } from 'history';
import { routerMiddleware } from 'react-router-redux';
import { applyMiddleware, createStore, compose } from 'redux';
import { responsiveStoreEnhancer } from 'redux-responsive';
import createDebounce from 'redux-debounced';
import thunk from 'redux-thunk';
import appMiddleware from './middleware/App';
import rootReducer from './reducers';

/**
 * redux devtools implementation.
 * We want to compose our enhancers normally within the production
 * environment (build), otherwise we want to compose our ehancers
 * with the devtools extension.
 */
export const composeFunc = () => {
  return process.env.NODE_ENV === 'production'
    ? compose
    : (() => {
      const { composeWithDevTools } = require('redux-devtools-extension');
      return composeWithDevTools;
    })();
};

/**
 * Third party redux middleware
 * - redux-thunk
 * - router
 * - react-router-redux
 * - history
 * - redux-responsive
 */
const libEnhancers = [
  thunk,
  routerMiddleware(createBrowserHistory({ basename: '/' })),
  createDebounce(),
].map(mw => applyMiddleware(mw));
libEnhancers.push(responsiveStoreEnhancer);

/**
 * Custom middleware
 * - appMiddleware
 */
const customEnhancers = [
  appMiddleware,
].map(mw => applyMiddleware(mw));

// compose enhancers with composeFunc
// https://redux.js.org/api/compose
const composedEnhancers = composeFunc().apply(null, [...libEnhancers, ...customEnhancers]);

// create store
// https://redux.js.org/api/createstore
const store = createStore(rootReducer, {}, composedEnhancers);

export default store;
