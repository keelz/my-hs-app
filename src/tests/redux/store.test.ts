import store from '../../redux/store';

describe('redux store', () => {
  it('returns a store correctly', () => {
    expect(store).not.toBeFalsy();
  });

  it('composes the default development state correctly', () => {
    const state = store.getState();
    expect(state).toMatchSnapshot();
  });
});
