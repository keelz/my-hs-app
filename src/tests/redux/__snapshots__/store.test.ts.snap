// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux store composes the default development state correctly 1`] = `
Object {
  "App": Object {
    "loaded": false,
  },
  "Browser": Object {
    "_responsiveState": true,
    "breakpoints": Object {
      "extraSmall": 480,
      "infinity": Infinity,
      "large": 1200,
      "medium": 992,
      "small": 768,
    },
    "greaterThan": Object {
      "extraSmall": true,
      "infinity": false,
      "large": true,
      "medium": true,
      "small": true,
    },
    "is": Object {
      "extraSmall": false,
      "infinity": true,
      "large": false,
      "medium": false,
      "small": false,
    },
    "lessThan": Object {
      "extraSmall": false,
      "infinity": false,
      "large": false,
      "medium": false,
      "small": false,
    },
    "mediaType": "infinity",
    "orientation": null,
  },
  "Cards": Object {
    "activeClassName": "DRUID",
    "data": Array [],
  },
}
`;
